### DISCUSSION TITLE
Invalidate auth cookies when password is changed

### DISCUSSION THREAD

**Discussion by g012** [keystonejs/keystone#7484] [Questions] [answered]
Hello,

I would like the stateless session cookies to fail when the user has changed her password. This would provide an easy way to disable sessions remotely when the user can't access the computer easily.

It's in the spirit of a comment of this SO question : 
https://stackoverflow.com/questions/21978658/invalidating-json-web-tokens

> A common approach for invalidating tokens when a user changes their password is to sign the token with a hash of their password. Thus if the password changes, any previous tokens automatically fail to verify. You can extend this to logout by including a last-logout-time in the user's record and using a combination of the last-logout-time and password hash to sign the token. This requires a DB lookup each time you need to verify the token signature, but presumably you're looking up the user anyway. – Travis Terry

Thanks !

**Comment by g012** [comment]
I think a possible solution would be a hook, triggered after session.get and user data loaded, which receives these two values. It would let us do whatever check we need, such as a disabled or restricted/banned account : I could create a random field `challenge` in the user list, and store it in the session data, then update the database value each time the user password is changed with a list or field hook ; then upon any further request, in the new hook, if session.challenge does not match user.challenge, the hook function would throw "Expired Token". Likewise if user.disabled it would throw "Account Disabled", ...

I can do that now I presume by wrapping all my session validations in list/field hooks into a function which would throw if it does not pass, but it's all over the place. This new hook would make it more tidy and less error-prone by centralizing this processing.

**Reply by g012** [reply]
I've actually tried this, and it doesn't work : after peeping at auth source, it seems it merges the user data it got from the database into the decrypted session object, hence loosing any duplicated field. This spot seems like the good place to put the validateSession(session, data) hook though, just before overwriting session fields with user db data : auth src / index.ts > createAuth() > withItemData() > get().

**Comment by borisno2** [comment, marked as answer]
Hi @g012, thanks for your question. This is possible by customising `get` and `start` in the session API.

Essentially you can make your session look something like [this](https://github.com/keystonejs/keystone/blob/047a84262b559dda3bd73d87e0bfd257182bde29/examples/custom-session-validation/keystone.ts#L34-L70), which will add a `startTime` to `session` and then check that `startTime` against `passwordChangedAt` (which you will need to add).

https://github.com/keystonejs/keystone/blob/047a84262b559dda3bd73d87e0bfd257182bde29/examples/custom-session-validation/keystone.ts#L34-L70

The `password` and `passwordChangedAt` can then look something like this, with a hook on the password field when changed

https://github.com/keystonejs/keystone/blob/047a84262b559dda3bd73d87e0bfd257182bde29/examples/custom-session-validation/schema.ts#L29-L57

A complete example has been added [here](https://github.com/keystonejs/keystone/blob/main/examples/custom-session-validation/keystone.ts), and should hopefully be an up-to-date reference.

- https://github.com/keystonejs/keystone/pull/7687

**Comment by g012** [comment]
Thanks a lot Josh, it's awesome !

Le lun. 27 juin 2022 à 01:43, Josh Calder ***@***.***> a
écrit :
